add support for scryer-prolog

Skgland opened the pull request:
This add the option of running plwm with scryer-prolog instead of swi.

> [!warning]
> The basics should be working at this stage, see the todo and help wanted section below for what is missing.
> When testing I recommend using a VM or at least make sure you are able to change back to a working window manager in case this doesn't.
> I.e. be able to switch tty, log into the terminal, and kill scryer-plwm (may need to kill display manager as well).

This splits the plwm binary into plwm-swi and plwm-scryer (the later is actually a script).

## Installation on debian based system

### Prerequisites 
- rust toolchain <https://www.rust-lang.org/tools/install>
- cargo-deb `cargo install cargo-deb`
- debian packages `run0 apt install libxft-dev libx11-dev libxrandr-dev`
- things that I am forgetting

### Install scryer-prolog

#### From CI Artifacts

From https://github.com/mthom/scryer-prolog/actions/runs/18056647934#artifacts

1. Download the scryer-prolog_<target> Artifact Zip
2. Install Scryer
   a. place the `scryer-prolog[.exe]` binary from the release folder somewhere on your `PATH`
   b. install the `scryer-prolog_<version>-1_<arch>.deb` package from the debian folder e.g. `apt install scryer-prolog_<version>-1_<arch>.deb`

#### Build from Source

##### Debian Package

```
git clone https://github.com/mthom/scryer-prolog.git
cd scryer-prolog
cargo deb
run0 apt install ./target/debian/*.deb
```

##### Manual

```
git clone https://github.com/mthom/scryer-prolog.git
cd scryer-prolog
cargo build --release
```

Then place ./target/release/scryer-prolog[.exe] somewhere on your PATH

### Build & Install plwm-{core,swi,scryer}

```
git clone https://github.com/Skgland/plwm.git --branch scryer
cd plwm
make install-debs
```
  
## Help Wanted

### Try everything and find out whats broken

I am not used to tiling window manager and plwm in general.
So I am having a hard time testing things as at this point the bugs are often more subtle and I often don't known if some behavior is due to 
- a bug in the prolog port
- I am not using it correctly
- a bug already present in stock swi plwm

Also using features that I haven't tried or don't know how to use would be helpful as this might 
reveal more predicates that are assumed present from swi but that are missing in scryer and havent been implemented/stubbed in the port.

### Signal handling

This requires some work in scryer-prolog as it currently doesn't have a way to handle signals besides hardcoded Ctr+C handling.

Some notes:
- The signal handler function itself should likely be relatively small i.e. check the signal configuration and abort the process, ignore the signal, flag the signal as received to the machine.
- In the case a custom predicate is to be evaluated on receiving a signal the machine needs to be adjusted to monitor these flags and cause the goal to be evaluated.
- Can this be done while scryer-prolog is still single threaded? How would that work while the machine is stuck evaluating an ffi call?

### Run Tests (SWI-prolog)

I would like some help with `set + (defaults can be set)` as that test appears to pass on main but fails here in a weird way.

Note: a false has been inserted into the test to stop it from breaking other tests that needs to be removed when trying to search for the root cause.

`warn_invalid_arg` and `global_key_newvalue - (nonassoc_g)` appear to also fail on master so I will ignore them as pre-existing failures.

### Command Line Argument Handling

Two ways to go about this:
1. contribute a command-line parsing module to scryer-prolog and use it here (idealy compatible with swi-prolog)
2. implement the required parsing in the scryer stubs here.


### Improve stub files 

I think this is best discussed in #113

## TODO

- [x] implement remaining function bindings from src/swi/plx.c in src/scryer/plx.pl
- [x] adjust install/uninstall scripts
- [ ] use mthom/scryer-prolog#3111 to correctly map the X11 API
  - requires a new scryer-prolog release, don't want to break support with scryer-prolog 0.10.0 until then

Skgland committed:
declare 0-arg predicates as atom :- ... . instead of atom() :- ... .

Skgland committed:
replace ceil with ceiling

Skgland committed:
fix invalid arguments 6.3.3.1

Skgland committed:
stub some things for now and import lists and iso_ext modules

Skgland committed:
adjust obt_arguments stub

Skgland committed:
adjust dynamic declarations

- Note: jobs/1 is no longer explicitly declared as shared, but based on swipl documentation this is the default anyways

Skgland committed:
move reassert definition to utils

Skgland committed:
define dynamic predicates also in settings

Skgland committed:
fix condig.pl

Skgland committed:
fix config loading

Skgland committed:
stub more things

Skgland committed:
start implementing ffi

Skgland committed:
implement more ffi

Skgland committed:
import lib assoc in utils

Skgland committed:
add two more C macro wrappers

Skgland committed:
adjust stubs

Skgland committed:
make running this easier `make run`, implement more ffi and make some usages of format compatible

Skgland committed:
assert that the Length of ArrayValues matches NElements

Skgland committed:
fix ffi function declarations

Bool = i32 and most functions have an unused i32 return value

Skgland committed:
implement x_create_simple_window

Skgland committed:
add script for running

Skgland committed:
ignore return value

Skgland committed:
use chars_si

as suggested by triska

Skgland committed:
implement x_ungrab_key/4

Skgland committed:
use chars_utf8bytes

Skgland committed:
stub atom_string/2

Skgland committed:
replace string/1 with chars_si/1

Skgland committed:
fix mixup of ModAtoms and Mods

also fix format usage

Skgland committed:
replace string/1 with chars_si/1

Skgland committed:
fix format usages

Skgland committed:
unfix grab_keys

Skgland committed:
implement x_string_to_keysym/2 and x_keysym_to_keycode/3

Skgland committed:
improve atom_string stub and add ignore stub

Skgland committed:
fix use of format/2 instead of format_//2

Skgland committed:
remove unecessary once

Skgland committed:
implement more ffi bindings

and fix xrr_query_extension

Skgland committed:
un-duplicate dynamic declaration, add more stubs

Skgland committed:
stub selectchk

Skgland committed:
implement x_utf8_text_list_to_text_property/5

Skgland committed:
fix incorrect ffi definition

Skgland committed:
use iso_ext:bb_put/bb_get

Skgland committed:
fix x_utf8_text_list_to_text_property

Skgland committed:
cleanup after map_ffi_arg fix

Skgland committed:
implement x_set_text_property/4

Skgland committed:
fix setup_fifo

Skgland committed:
implement c_free/1 and x_next_event/2

Skgland committed:
order events by event id and fix incorrect id for buttonpress

Skgland committed:
apply suggestions by triska

Skgland committed:
remove use_foreign_library stub

Skgland committed:
declare ignore/1 as a meta_predicate

Skgland committed:
use list_si for is_list

anagram

Skgland committed:
remove declaration of unused dynamic predicate globals/2

Skgland committed:
add dumy rust crate so that cargo-deb can be used to create a .deb package

Skgland committed:
fix missing symlink

Skgland committed:
resolve symlinks before trying to get the directory

Skgland committed:
add maintainer script to register alternatives

Skgland committed:
fix setting LD_LIBRARY_PATH

Skgland committed:
don't try to use stubs from config

Skgland committed:
add error context

Skgland committed:
fix argument order

Skgland committed:
fix wrong predicate name

Skgland committed:
don't module qualify ffi:with_locals

Skgland committed:
fix swapped ffi:read_ptr arguments

Skgland committed:
log eventloop crash

Skgland committed:
fix decode_event

Skgland committed:
remove extra space in atom names

Skgland committed:
debug eventloop

Skgland committed:
debug x_next_event

Skgland committed:
[WIP] swi & scryer compat at the same time

Skgland committed:
apply triskas suggestion

https://github.com/Skgland/plwm/commit/ad2a61b7f86d567d6ef6caa61ddd69000254802b#r164282133

Skgland committed:
split plwm deb into plwm-core and plwm-scryer and add plwm-swi

Skgland committed:
add swi patch

Skgland committed:
fix newer swi

Skgland committed:
adjust debs

Skgland committed:
adjust lib install path

Skgland committed:
remove need for swi.patch

Skgland committed:
[WIP] use goal_expansion

Skgland committed:
try to reduce the diff

Skgland committed:
fix recursive forall goal_expansion

Skgland committed:
fix missing sub_string/5 and format/2

Skgland committed:
add debs target

Skgland committed:
fix SWIFLAGS having a random none

Skgland committed:
fix swi compat

Skgland committed:
fix swi some more

Skgland committed:
more fixes and diff reduction

Skgland committed:
fix call_with_error_context in swi

Skgland committed:
add terminal emulater shortcut

Skgland committed:
add a dependency on suckless-tools package

Skgland committed:
fix call_with_error_context in scryer

Skgland committed:
remove accidentally commited diff.txt

Skgland committed:
fix recursive goal_expansion

presumably

Skgland committed:
add install-debs target

Skgland committed:
ensue writeln/1 and writeln/2 are considered current predicates

Skgland committed:
ensure goal is module qualified

Skgland committed:
apply suggestion by triska

https://github.com/Skgland/plwm/commit/750696ab2b9ad6a04beec94f3791cbe1bfe8119c#r164123169

Skgland committed:
ensure unsupported event is handled correctly when using scryer

Skgland committed:
fix missing parenthesis

Skgland committed:
fix incorrect argument count

Skgland committed:
debug decode_event(motionnotify,_,_)

Skgland committed:
fix missing time field

Skgland committed:
define max_member/3 when using scryer

Skgland committed:
module qualify max_member_helper

Skgland committed:
module qualify ignore/1 Goal

Skgland committed:
handle more swi specific predicates in scryer

Skgland committed:
import lists library module

Skgland committed:
module qualify char_type_helper/2

Skgland committed:
fix error when Cat = white

Skgland committed:
handle shell/1 and shell/2 in scryer

triska commented:
Thank you a lot for this incredible work!

It also needs a few development headers installed, mentioned in the plwm README:

<pre>
<b>$ sudo apt-get install libxft-dev libx11-dev libxrandr-dev</b>
</pre>

Skgland commented:
I'm using a default install with gnome. I don't know why this would be the case with xfce4.

Skgland commented:
You could try editing the plwm-scryer script to redirect the stdout/stderr output of scryer-prolog into a file.

Skgland committed:
support delete/3 in scryer

Skgland committed:
implement x_get_window_attributes/4 for scryer

Skgland committed:
fix XWindowAttributes initializer argument count

triska commented:
For the next few days I only have access to a Ubuntu system, and on Ubuntu the installation from deb files does not work:

<pre>
<b>$ sudo apt install ./target/debian/*.deb</b>
...
Preparing to unpack .../scryer-prolog_0.9.4-1_amd64.deb ...
Unpacking scryer-prolog (0.9.4-1) ...
Setting up scryer-prolog (0.9.4-1) ...
<b>N: Download is performed unsandboxed as root as file '.../scryer-prolog/target/debian/scryer-prolog_0.9.4-1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)</b>
</pre>

Is there anything I can do manually to try it also on Ubuntu? A Scryer executable is available in `~/scryer-prolog/target/release/scryer-prolog`.

The **`plwm-scryer`** executable built successfully and is available in `~/plwm/src/plwm-srcyer`.

Skgland committed:
implement x_get_class_hint/4 for scryer

Skgland commented:
The part of the log you shared doesn't actually include an error.
It is just a warning/information that instead of a sandboxed install using the user `_apt` an unsandboxed install was performed as `root` as the .deb wasn't accessible by `_apt`.

Skgland commented:
Also I think I had a brain-fart yesterday when I opened this draft. 
I must have mixed up plwm-swi and plwm-scryer after I finished getting the former working again and the later to the state before starting to restore swi functionality.
So this isn't actually as far along as I stated.

Skgland commented:
I would expect the following steps to get it running without installing the debs:

For plwm-scryer:

- placing the scryer-prolog binary (or a symlink to it) on the PATH
- placing a symlink to `<git-checkout>/src/plwm-scryer` on the PATH
  - the relative location of plwm-scryer to the other files in `src/**` is important
- placing plwm-scryer.desktop in `/usr/share/xsessions/`
- building and placing `<git-checkout>/bin/x11plwm.so` (or a symlink to it) in `/usr/lib/`
- `apt install subckless-tools`

For plwm-swi

- place plwm-swi (or a symlink to it) on the PATH
- placing plwm-swi.desktop in `/usr/share/xsessions/`
- building and placing `<git-checkout>/bin/plx.so` (or a symlink to it) in `/usr/local/lib/`
  - might not be even necessary if plx.so is compiled into plwm-swi?
- `apt install subckless-tools`

Skgland committed:
fix missing parenthesis

Skgland committed:
remove exports for scryer plx module

the predicates are always used module qualified

Skgland committed:
implement x_close_display/1

Skgland committed:
implement remaoning plx.c bindings in plx.pl

Skgland committed:
fix incorrect predicate name

Skgland committed:
fix missing argument

Skgland committed:
handle term_to_atom/2

Skgland committed:
fix misspelling

Skgland committed:
handle nb_current

Skgland committed:
module qualify active_mon_ws

Skgland committed:
module qualify win_properties

Skgland committed:
module qualify border_width

Skgland committed:
module qualify win_newproperties

Skgland committed:
define foreign struct XConfigureEvent

Skgland committed:
module qualify monitors

Skgland committed:
module qualify monws_keys

Skgland committed:
defined 'XGetTextProperty'/5

Skgland committed:
handle nb_delete/1

Skgland committed:
module qualify inner_gaps

Skgland committed:
module qualify update_ws_atoms

Skgland committed:
handle between/3

Skgland committed:
import between

Skgland committed:
fix hang

Skgland committed:
handle nextto/3

Skgland commented:
Now it should actually be at (or even a bit beyond) the state that I thought this was in when opening this Draft PR.

Skgland commented:
> Ah yes, **awesome**, haha! Findings so far:
> 
> `Alt + a` yields a lookahead search box at the top of the screen that successfully completes `terminal` to **`gnome-terminal`**. However, when I confirm with `RET` no terminal opens. 
> 

x-terminal-emulator (the same as Ctrl+Alt+t was working for me)

> `Ctrl-Alt+t` successfully opens a terminal that displays a bright cursor. However, no input seems to be accepted, I cannot type anything into the terminal. I can successfully open further terminals with `Ctrl+Alt+t`.

I don't think new windows get focused automatically, toggling full screen worked for me to get it focused

> 
> The string `plwm` still does not appear in the output of `journalctrl -e`.
> 

I noticed that I don't get output when I start it with startx on a different tty, so this might be related. I will try adding a redirect of stdout/stderr to a file to the plwm-scryer script so that at least that can be used to retrieve logs.

Skgland commented:
Though I might not get to it today as I am visiting my parents to setup a printer 🖨️.

triska commented:
Please greet your parents and tell them about the international interest `plwm` and your work attracts! People from Austria, Brazil, Canada, the US and other countries watch the current developments in awe and look forward to using Scryer-powered `plwm` and the presentation at the meetup!

Skgland commented:
> `journalctrl -e`
There appears to be an extra r, it should be `journalctl -e`

Skgland committed:
log to $XDG_STATE_HOME/plwm/plwm-scryer.log

Skgland committed:
append to log file rather than truncate it

Skgland committed:
remove some no longer necessary debug output

Skgland committed:
fix misspelled predicate

triska commented:
Latest news from a **Debian** system (Trixie):

First, as you already suspected, I get an installation issue also on Debian (even though I followed the shown steps):

<pre>
<b>$ make install-debs</b>
...
Preparing to unpack .../plwm-scryer_0.1.0-1_amd64.deb ...
Unpacking plwm-scryer (0.1.0-1) over (0.1.0-1) ...
Setting up plwm-core (0.1.0-1) ...
Setting up plwm-scryer (0.1.0-1) ...
<b>Notice: Download is performed unsandboxed as root as file '~/plwm/target/debian/plwm-core_0.1.0-1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)</b>
</pre>

I worked around this issue like I did on Ubuntu, by `chown`-ing and copying the created `*.deb` files to `/tmp/` (which is owned by root) and installing them from there.

As before, I can select `plwm-scryer` when logging in (using the WM selection menu of XFCE), but on this system, **I get thrown back immediately to the login screen**. This is different from what I saw on Ubuntu (where I had Gnome installed).

The new logging facility will be very helpful for such situations, thank you a lot!

In my case, I get:

<pre>
<b>$ cat plwm-scryer.log</b> 
   error(existence_error(procedure,array_type/3),array_type/3).
main causes: error(existence_error(procedure,on_signal/3),on_signal/3)
</pre>

Skgland commented:
> Latest news from a **Debian** system (Trixie):
> 
> First, as you already suspected, I get an installation issue also on Debian (even though I followed the shown steps):
> 
> <pre>
> <b>$ make install-debs</b>
> ...
> Preparing to unpack .../plwm-scryer_0.1.0-1_amd64.deb ...
> Unpacking plwm-scryer (0.1.0-1) over (0.1.0-1) ...
> Setting up plwm-core (0.1.0-1) ...
> Setting up plwm-scryer (0.1.0-1) ...
> <b>Notice: Download is performed unsandboxed as root as file '~/plwm/target/debian/plwm-core_0.1.0-1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)</b>
> </pre>
> 
> I worked around this issue like I did on Ubuntu, by `chown`-ing and copying the created `*.deb` files to `/tmp/` (which is owned by root) and installing them from there.

I am getting the same notice and am just ignoring it. It's neither an error not a warning and the installation should succeed and work the same anyway.

> 
> As before, I can select `plwm-scryer` when logging in (using the WM selection menu of XFCE), but on this system, **I get thrown back immediately to the login screen**. This is different from what I saw on Ubuntu (where I had Gnome installed).
> 
> The new logging facility will be very helpful for such situations, thank you a lot!
> 
> In my case, I get:
> 
> <pre>
> <b>$ cat plwm-scryer.log</b> 
>    error(existence_error(procedure,array_type/3),array_type/3).
> main causes: error(existence_error(procedure,on_signal/3),on_signal/3)
> </pre>

This makes me suspect that an older scryer-prolog version without the changes from my ffi++ branch is being used. Did you have such a version installed and could it still be on the PATH and be prioritised over the in installed via the .deb file?

Skgland committed:
log which scryer-prolog is used

Skgland committed:
append to log rather than  truncate

Skgland committed:
fix is dir check

Skgland committed:
implement max_member/2 and max_list/2

Skgland committed:
replace ~p with ~q as scryer does not have the former

Skgland committed:
debug handle_event

- might break swi

Skgland committed:
change writeln_helper to use ~q instead of ~s

Skgland committed:
module qualify keymaps

Skgland committed:
use write_term with option double_quotes(true) for writeln_helper

Skgland committed:
fix missing tilde

Skgland committed:
debug Action instead of handle_event

Skgland committed:
handle string_length/2

Skgland committed:
fix missing tilde

Skgland committed:
module qualify menucmd

Skgland committed:
import library(process)

Seeker04 commented on Cargo.toml:
You generate a Debian package using `cargo` and `cargo-deb`. Is it usual for non-Rust projects to do this? I'm not necessarily against it, just curious. (scryer-prolog is only a binary dependency for us, if I understand correctly).

On a more general note: I feel like packaging and dependency management should go to the respective distros' repos instead of here upstream. Of course, for now we can keep this to have an easy way to install on Debian-based systems. It's definitely handy. However, I'd separate it from the main build logic. That is, from the Makefile I'd remove the `deb*` rules and `install-debs` and put those in a separate script or Makefile e.g. inside `tools/extra/`, if it's possible. Also, there's `tools/install.sh` which aims to be a distro-agnostic way to install. All it lacks is pulling packages dependencies as that's distro specific and should be the responsibility of downstream packagers.

Skgland replied:
> You generate a Debian package using cargo and cargo-deb. Is it usual for non-Rust projects to do this? I'm not necessarily against it, just curious. (scryer-prolog is only a binary dependency for us, if I understand correctly).

I just find it the easiest way to generate debian packages. It allows building debian packages on non-debian system (assuming cross-compilation is setup properly even on windows) and I find it way less cryptic/convoluted than debians own package build system.

> On a more general note: I feel like packaging and dependency management should go to the respective distros' repos instead of here upstream. Of course, for now we can keep this to have an easy way to install on Debian-based systems. It's definitely handy. However, I'd separate it from the main build logic. That is, from the Makefile I'd remove the deb* rules and install-debs and put those in a separate script or Makefile e.g. inside tools/extra/, if it's possible. Also, there's tools/install.sh which aims to be a distro-agnostic way to install. All it lacks is pulling packages dependencies as that's distro specific and should be the responsibility of downstream packagers.

While having an install script is nice if a package isn't available. I strongly prefer a package and have it keep track of installed files as well as dependencies.
Though I don't expect/intend this package to be directly used downsteam by distro maintainers.

Seeker04 commented on symlinks/plwm-scryer:
I don't understand what purpose this link serves. I see that `Cargo.toml` uses it. Instead of storing in git, could we only generate this if the user specifically runs a Debian install (i.e. from `tools/extra/` in tandem with my previous comment)?

Skgland replied:
For now I have moved the symlink into tools/extra, but kept it in git.
An option would be to add a Makefile target for creating the symlink and have the `deb-scryer` target depend on it. Ideally the cargo-deb issue liked above would be fixed and we could just get rid of the workaround.

Seeker04 commented on config/config.pl:
```diff
@@ -155,75 +155,75 @@
-  super + d                 ->  change_nmaster(-1)              ,
-  super + h                 ->  change_mfact(-0.05)             ,
-  super + l                 ->  change_mfact(+0.05)             ,
+  (super +         j         ->  shift_focus(down))               ,
```
Does scryer need the parentheses for the keymaps, hooks, etc.? Is it an error or warning that warrants these? (It's fine if its needed, just curious).

triska replied:
Yes, these parentheses are needed for *all* strictly conforming Prolog systems, adding them is good for portability also with future&nbsp;systems that are candidates for a `plwm`&nbsp;port. Scryer correctly rejects invalid syntax and helps to quickly detect such portability issues.

GNU&nbsp;Prolog is also an excellent free reference system that reliably tells us what is Prolog&nbsp;syntax and what is not:

<pre>
<b>| ?- T = [a -> b, c].</b>   
uncaught exception: error(syntax_error('user_input:1 (char:44) , | ] or operator expected in list'),read_term/3)

<b>| ?- T = [(a -> b), c].</b>
T = [(a->b),c]
</pre>

Please also see https://github.com/mthom/scryer-prolog/discussions/3030 for more information.

Seeker04 replied:
It makes sense now. Thank you for the detailed explanation!

Seeker04 commented on plwm-scryer.desktop:
Thank you for adding the .desktop files! It will solve #70. I think, we could put these under `config/` and make `tools/install.sh` install them.

Seeker04 commented on src/utils.pl:
```diff
@@ -296,3 +299,22 @@
 str_withoutlastch(Str, NewStr) :- string_length(Str, Len), Lenm1 is Len - 1, sub_string(Str, 0, Lenm1, _, NewStr).
 
+
+%! reassert(++Callable:callable) is det
```
Previously, when I put this under the `utils` module, it didn't work, because then swi asserted everything with the `utils:` prefix. Was this resolved somehow?

Skgland replied:
No, I don't think this is resolved, I think I just didn't yet run into problems due to this.

triska commented on src/plwm.pl:
```diff
@@ -1399,25 +1383,27 @@
+handle_event_(keypress, [_, _, _, _, _, _, _, _, _, _, _, State, Keycode, _]) :-
 	(keymap_internal(Keycode, State, Action) ->
-		catch(ignore(Action), Ex, (writeln(Ex), true))
+		catch(ignore($Action), Ex, (writeln(Ex), true))
```
This is from Scryer's `library(debug)`, it emits a trace of the executed goal:

https://github.com/mthom/scryer-prolog/blob/dc05526c71169a88f90ba91bbbb72cad450782b0/src/lib/debug.pl#L47

Seeker04 replied:
> This is from Scryer's library(debug), it emits a trace of the executed goal

Tracing these goals sounds super useful! It's always been a hassle to find out where goal failure propagated from.

> Note that swi-prolog has a built-in $/[1](https://github.com/Seeker04/plwm/pull/114#user-content-fn-1-7c856981da65138332d9e583ba789164)1 which does something different.
Which likely explains the errors you are seeing when using swi (mentioned in https://github.com/Seeker04/plwm/pull/114#issuecomment-3289526276).

If I add `:- set_prolog_flag(determinism_error, silent).` to `src/swi/init.pl`, these warnings are suppressed.

Seeker04 commented:
@Skgland  First and foremost, sorry for taking so long to react. Thank you very much for this amazing work! It looks very promising. I never thought progress towards scryer could be made this fast!

I quickly glanced over the changes and made some comments above. I understand, this is still heavily in a draft stage, but early feedback might be useful.

I'm still yet to fire up the scryer version, because I'm not on a Debian-based system, but I'll tackle it.

Cheers

Skgland committed:
move desktop file into config folder

Skgland committed:
load init.pl for tests

Skgland committed:
address some review comments

triska commented on src/plwm.pl:
```diff
@@ -1649,8 +1635,7 @@
 				global_key_newvalue(monitor_geom, Output, Geom),
-				format(string(Msg), "Monitor \"~s\" geometry reconfigured", [Output]),
-				writeln(Msg)
+			compat_format("Monitor \"~s\" geometry reconfigured~n", [Output])
```
Regarding `compat_format/2`: If a system provides `format(chars(...), ...)`, then it  is easy to describe **`format_//2`** in terms of that:

<pre>
format_(Fs, Args) --&gt; call(format__(Fs, Args)).
format__(Fs, Args, Cs0, Cs) :- format(chars(Cs0, Cs), Fs, Args).
</pre>

With this compatibility definition, the Scryer code (using `format_//2`) can work in multiple systems.

Skgland committed:
adjust install/uninstall script

Skgland commented on Makefile:
```diff
@@ -11,25 +11,53 @@ OFLAGS  = -O2
 #================================== Build =====================================
 
-$(PLWM): src/*.pl $(PLX_SO)
+run-scryer: $(X11PLWM_SO)
```
Yeah, the `run-*` targets don't make too much sense. They were useful when everything was broken to the point the even just loading failed, but now that error actually happen at runtime they don't serve much purpose anymore.

I am not working much with make so I didn't take into account that the first defined target is the default target, ups. 

I don't think it makes sense to install `src/plwm-scryer` in `/usr/local/bin` as it would require all the other module files to be places next to it to work properly.

Skgland committed:
fix path to debian packages

Skgland commented on src/scryer/x11plwm.c:
Sounds reasonable. 
Would it make sense to just have `scryer/x11plwm.c` and `swi/x11plwm.c` instead of `scryer/x11plwm-scryer.c` and `swi/x11plwm-swi.c` while having them still compile to `x11plwm-scryer.so` and `x11plwm-swi.so`,
to not have a duplicate swi/scryer in the path or would it be better to have matching filenames between the .c and .so file?

Skgland replied:
Ah, a problem with this is that this would change the module name from `plx` to `plx-swi`/`x11plwm-swi` unless I am missing a way to define under what module name `use_foreign_library` loads the predicates exposed by the dynamic library.

Seeker04 replied:
Then we can also go with:
```
swi/plx.c
scryer/plx.c
```
I did some search and haven't found any other lib called `plx.so`. Still, if someone reports a name conflict, we can rename it any time...

Skgland committed:
revert change from format/3 and writeln/1 to format/2

Skgland committed:
fix logo path in README.md

Skgland committed:
move reassert back into plwm.pl

Skgland committed:
inclue usr/local/lib in  LD_LIBRARY_PATH

Skgland committed:
try to fix add

Skgland committed:
remove unused Makefile

triska commented on src/setting.pl:
```diff
@@ -317,7 +317,7 @@
 add(Setting, Value) :-
 	member(Setting, [workspaces, layout_default_overrides, menucmd, keymaps, rules, hooks]),
-	call(Setting, PrevList),
+	call(user:Setting, PrevList),
```
That, or invoke `call/2` in `user` with `user:call(... ,...)`. 

(?)

Skgland committed:
fix some tests

- a few tests have a false added to force them to fail as the otherwise they currently break plunit and cause all later tests to not run

Skgland committed:
fix list_cmds

Skgland committed:
fix string_concat

Skgland committed:
hanlde path aliases in process_create in scryer

Skgland committed:
fix unknown predicate (:)/2 in scryer

Skgland committed:
handle read_string/3 in scryer

Skgland committed:
handle split_string/4 in scryer

Skgland committed:
add more mocks

don't know why this was working without these stubs in the first place

Skgland committed:
fix utils:global_key_value mock not working

Skgland committed:
fix update_all_borders

Skgland committed:
acual fix

Skgland commented:
Update on the state of test suite results for plwm-swi compared to main plwm

<details>
<summary>Output of `make test` on main</summary>

```
~/git/plwm$ make test
tests/run_unit_tests.sh
tests/unit_tests/animation.plt
% PL-Unit: animation_tests ................ done
% All 16 tests passed
tests/unit_tests/fifo.plt
% PL-Unit: fifo_tests ....error(syntax_error(end_of_file),file(/tmp/test-terms,1,4,3))
. done
% All 5 tests passed
tests/unit_tests/layout.plt
% PL-Unit: layout_tests ........... done
% All 11 tests passed
tests/unit_tests/menu.plt
% PL-Unit: menu_tests ......................_4244.................
Warning: Goal (retract(menu:menucmd(_10678)),retract(menu:set(_10692,_10694))) failed: '(cleanup handler)'
.
Warning: Goal (retract(menu:menucmd(_246)),retract(menu:set(_260,_262))) failed: '(cleanup handler)'
.
Warning: Goal (retract(menu:menucmd(_1492)),retract(menu:add(_1506,_1508))) failed: '(cleanup handler)'
.
Warning: Goal (retract(menu:menucmd(_3008)),retract(menu:add(_3022,_3024))) failed: '(cleanup handler)'
.
Warning: Goal (retract(menu:menucmd(_4458)),retract(menu:dump_settings(_4472,_4474))) failed: '(cleanup handler)'
.
Warning: Goal (retract(menu:menucmd(_5948)),retract(menu:dump_settings(_5962,_5964))) failed: '(cleanup handler)'
..... done
% All 49 tests passed
tests/unit_tests/setting.plt
% PL-Unit: setting_tests ....warning: invalid value: _33590 for setting: foo, ignored
.warning: invalid value: foo for setting: default_nmaster, ignored
.....warning: invalid value: _39794 for setting: foo, ignored
.warning: invalid value: foo for setting: default_nmaster, ignored
....warning: invalid value: [a,b,c,1.0] for setting: workspaces, ignored
warning: invalid value: [(_46220,a->3,0.5,grid),("Mon1",_46258->2,0.7,rmaster),(_46298,_46300->_46304,_46310,_46312),foo] for setting: layout_default_overrides, ignored
warning: invalid value: ["dmenu","-i","-l",42] for setting: menucmd, ignored
warning: invalid value: [(super+q->close_focused),(super+shift+q->quit),("AudioRaiseVolume"->shellcmd("pulseaudio-ctl up")),(foo+bar+z->writeln("foo"))] for setting: keymaps, ignored
warning: invalid value: [("a","b","c"->"Mon1",ws1,fullscreen),(_46380,_46386,exact("c")->_46402,ws1,[left,top,1/2,1/3]),(_46462,_46468,_46470->"Mon1",_46486,managed),"bar"] for setting: rules, ignored
warning: invalid value: [(start->writeln("plwm starting")),(quit->writeln("plwm quitting")),(window_create_post->writeln("ws created")),baz] for setting: hooks, ignored
.warning: invalid value: foo for setting: default_nmaster, ignored
warning: invalid value: foo for setting: default_mfact, ignored
warning: invalid value: foo for setting: default_layout, ignored
warning: invalid value: foo for setting: attach_bottom, ignored
warning: invalid value: foo for setting: border_width, ignored
warning: invalid value: foo for setting: border_width_focused, ignored
warning: invalid value: foo for setting: border_color, ignored
warning: invalid value: foo for setting: border_color_focused, ignored
warning: invalid value: foo for setting: snap_threshold, ignored
warning: invalid value: foo for setting: outer_gaps, ignored
warning: invalid value: foo for setting: inner_gaps, ignored
warning: invalid value: foo for setting: workspaces, ignored
warning: invalid value: foo for setting: starting_workspace, ignored
warning: invalid value: foo for setting: hide_empty_workspaces, ignored
warning: invalid value: foo for setting: ws_format, ignored
warning: invalid value: foo for setting: ws_format_occupied, ignored
warning: invalid value: foo for setting: layout_default_overrides, ignored
warning: invalid value: foo for setting: bar_classes, ignored
warning: invalid value: foo for setting: bar_placement, ignored
warning: invalid value: foo for setting: fifo_enabled, ignored
warning: invalid value: foo for setting: fifo_path, ignored
warning: invalid value: foo for setting: menucmd, ignored
warning: invalid value: foo for setting: animation_enabled, ignored
warning: invalid value: foo for setting: animation_time, ignored
warning: invalid value: foo for setting: animation_granularity, ignored
warning: invalid value: foo for setting: modkey, ignored
warning: invalid value: foo for setting: scroll_up_action, ignored
warning: invalid value: foo for setting: scroll_down_action, ignored
warning: invalid value: foo for setting: keymaps, ignored
warning: invalid value: foo for setting: rules, ignored
warning: invalid value: foo for setting: hooks, ignored
.........warning: starting_workspace: 1 is not a workspace, defaulting to ws0
... done
% All 28 tests passed
tests/unit_tests/utils.plt
% PL-Unit: utils_tests ............
Warning: delete_file/1: Arguments are not sufficiently instantiated
% All 12 tests passed
ERROR: -g run_tests: arg/3: Arguments are not sufficiently instantiated
```

</details>

<details>
<summary>Output of `make test` on scryer</summary>

```
bennet@DESKTOP-RP6E64N:~/git/plwm$ make test
tests/run_unit_tests.sh
tests/unit_tests/animation.plt
% PL-Unit: animation_tests ................ done
% All 16 tests passed
tests/unit_tests/fifo.plt
% PL-Unit: fifo_tests ....error(syntax_error(end_of_file),file(/tmp/test-terms,1,4,3))
. done
% All 5 tests passed
tests/unit_tests/layout.plt
% PL-Unit: layout_tests ........... done
% All 11 tests passed
tests/unit_tests/menu.plt
% PL-Unit: menu_tests ......................_3048.................
Warning: Goal (retract(user:menucmd(_9486)),retract(user:set(_9500,_9502))) failed: '(cleanup handler)'
.
Warning: Goal (retract(user:menucmd(_256)),retract(user:set(_270,_272))) failed: '(cleanup handler)'
.
Warning: Goal (retract(user:menucmd(_1502)),retract(user:add(_1516,_1518))) failed: '(cleanup handler)'
.
Warning: Goal (retract(user:menucmd(_3018)),retract(user:add(_3032,_3034))) failed: '(cleanup handler)'
.
Warning: Goal (retract(user:menucmd(_4468)),retract(user:dump_settings(_4482,_4484))) failed: '(cleanup handler)'
.
Warning: Goal (retract(user:menucmd(_5958)),retract(user:dump_settings(_5972,_5974))) failed: '(cleanup handler)'
..... done
% All 49 tests passed
tests/unit_tests/setting.plt
Warning: /home/bennet/git/plwm/tests/unit_tests/setting.plt:39:
Warning:    Redefined static procedure utils:global_key_value/3
Warning:    Previously defined at /home/bennet/git/plwm/src/utils.pl:164
% PL-Unit: setting_tests ....warning: invalid value: _20594 for setting: foo, ignored
.warning: invalid value: foo for setting: default_nmaster, ignored
...
ERROR: 
    [[ EXCEPTION while printing message url('/home/bennet/git/plwm/tests/unit_tests/setting.plt':118)
       with arguments []:
       raised: type_error(text,url('/home/bennet/git/plwm/tests/unit_tests/setting.plt':118))
    ]]
:
        test set + (defaults can be set): failed

.warning: invalid value: _2126 for setting: foo, ignored
.warning: invalid value: foo for setting: default_nmaster, ignored
.....warning: invalid value: foo for setting: default_nmaster, ignored
warning: invalid value: foo for setting: default_mfact, ignored
warning: invalid value: foo for setting: default_layout, ignored
warning: invalid value: foo for setting: attach_bottom, ignored
warning: invalid value: foo for setting: border_width, ignored
warning: invalid value: foo for setting: border_width_focused, ignored
warning: invalid value: foo for setting: border_color, ignored
warning: invalid value: foo for setting: border_color_focused, ignored
warning: invalid value: foo for setting: snap_threshold, ignored
warning: invalid value: foo for setting: outer_gaps, ignored
warning: invalid value: foo for setting: inner_gaps, ignored
warning: invalid value: foo for setting: workspaces, ignored
warning: invalid value: foo for setting: starting_workspace, ignored
warning: invalid value: foo for setting: hide_empty_workspaces, ignored
warning: invalid value: foo for setting: ws_format, ignored
warning: invalid value: foo for setting: ws_format_occupied, ignored
warning: invalid value: foo for setting: layout_default_overrides, ignored
warning: invalid value: foo for setting: bar_classes, ignored
warning: invalid value: foo for setting: bar_placement, ignored
warning: invalid value: foo for setting: fifo_enabled, ignored
warning: invalid value: foo for setting: fifo_path, ignored
warning: invalid value: foo for setting: menucmd, ignored
warning: invalid value: foo for setting: animation_enabled, ignored
warning: invalid value: foo for setting: animation_time, ignored
warning: invalid value: foo for setting: animation_granularity, ignored
warning: invalid value: foo for setting: modkey, ignored
warning: invalid value: foo for setting: scroll_up_action, ignored
warning: invalid value: foo for setting: scroll_down_action, ignored
warning: invalid value: foo for setting: keymaps, ignored
warning: invalid value: foo for setting: rules, ignored
warning: invalid value: foo for setting: hooks, ignored
............ done
% 1 test failed
% 27 tests passed
ERROR: -g consult('tests/unit_tests/setting.plt'),run_tests: false
tests/unit_tests/utils.plt
% PL-Unit: utils_tests ............
Warning: delete_file/1: Arguments are not sufficiently instantiated
ERROR: 
    [[ EXCEPTION while printing message url('/home/bennet/git/plwm/tests/unit_tests/utils.plt':104)
       with arguments []:
       raised: type_error(text,url('/home/bennet/git/plwm/tests/unit_tests/utils.plt':104))
    ]]
:
        test warn_invalid_arg: failed

............
ERROR: 
    [[ EXCEPTION while printing message url('/home/bennet/git/plwm/tests/unit_tests/utils.plt':249)
       with arguments []:
       raised: type_error(text,url('/home/bennet/git/plwm/tests/unit_tests/utils.plt':249))
    ]]
:
        test global_key_newvalue - (nonassoc_g): failed

....................... done
% 2 tests failed
% 47 tests passed
ERROR: -g consult('tests/unit_tests/utils.plt'),run_tests: false
make: *** [Makefile:76: test] Error 1
```

</details>

Using 

```
swipl --version
SWI-Prolog version 8.4.2 for x86_64-linux
```

Summary 

| Test File | Pass | Fail |
|--------|--------|--------|
| animation.plt | 16 (±0) | 0 (±0) |
| layout.plt | 11 (±0) | 0 (±0) |
| menu.plt | 49 (±0) | 0 (±0) |
| setting.plt | 27 (-1) | 1 (+1) |
| utils.plt | 47 (+35) | 2 (+2) |

### setting.plt

- Test `set + (defaults can be set)` fails in a way that causes all following test to be skipped so a `false` has been inserted to make it fail early without bricking other tests. 

### utils.plt

It appears one of the tests is broken on main in a way that causes all following test to be skipped
and it appears I made a change fixing that 🤷 resulting in additional test being run.

- `warn_invalid_arg`
- `global_key_newvalue - (nonassoc_g)`

Skgland committed:
ensure windows global is an assoc without overriding utils:global_key_value

Skgland committed:
Merge branch 'main' into scryer

Seeker04 commented on tests/unit_tests/setting.plt:
```diff
@@ -329,6 +355,8 @@ test("geometry_spec -") :-
 
 test("update_all_borders", [
 	setup((
+		list_to_assoc([("Mon1"-ws1)-[1,2], ("Mon1"-ws2)-[3,4,5]], Assoc)
```
A comma is missing from the end of this line.

Skgland replied:
Mh, how/when did that happen? I am pretty sure I tested this, but that shouldn't be possible with this comma missing.

Skgland replied:
Looks like swi just manages to recover from that missing comma and continues to run all other test. Maybe I just missed the error message in the wall of text (though it being red should have stood out).
Anyway I pushed a commit to add that missing comma.

Skgland committed:
fix missing comma

Skgland commented:
> I get this error right after launching `plwm-scryer`:
> 
> ```
> /usr/local/bin/scryer-prolog
> "scryer/init.pl initialized"
> main causes: error(syntax_error(incomplete_reduction),read_term/3:188)
> ```
> 
> Rust and Scryer versions are up to date:
> 
> ```
> rustc --version
> rustc 1.91.0 (f8297e351 2025-10-28)
> 
> scryer-prolog --version
> v0.10.0-17-ge4d96925
> ```
> 
> Have you encountered this or do you have some idea where this might originate from?

Maybe I messed up the rebase/merge yesterday, I didn't test it as it was a rather trivial rebase/merge (I should have known better).
I don't get that error and without mthom/scryer-prolog#3133 we don't know which file is the offending one.

Skgland committed:
print a message before and after loading configs

so we can pin point errros during config loading

triska commented:
A Prolog syntax error in a file should also be detectable without launching plwm. For instance, can you successfully *consult* the configuration file (if any) with Scryer Prolog, by specifying it on the command&nbsp;line?

Seeker04 commented:
Oh, sorry. It was indeed my config that was faulty :S

Now `plwm-scryer` launces on my machine too and core window management functionality works! (tiling, floating windows, fullscreen, workspace switching, etc.)

I see some hiccups withs the hooks and menu operations. Also, sometimes the `quit` action hangs the process and I have to kill it (I have a hunch this also has to do with a stuck dmenu child process...)

I'll write a proper assessment of the issues I find.

Either way, the core window managment works in Scryer which is awesome! :champagne:
